Base field

mfrachet opened the pull request:
Storybook available on https://design-system-git-base-field-strapijs.vercel.app/?path=/story/field--default-story

These components aim to be building blocks for higher-level components such as `TextInput` but also the select, textarea, dropdown etc...

## APIs
Most simple case

```jsx
 <Field name="email">
  <Stack size={1}>
    <FieldLabel>Email</FieldLabel>
    <FieldInput type="text" placeholder="Placeholder" value={value} onChange={(e) => setText(e.target.value)} />
  </Stack>
</Field>
```

Most complex case

```jsx
 <Field name="password" hint="Description line" error={email.length > 5 ? 'Email is too short' : undefined}>
  <Stack size={1}>
    <Grid cols="auto auto 1fr" gap={1}>
      <FieldLabel>Password</FieldLabel>
      <Tooltip content="Content of the tooltip">
        <button
          style={{ border: 'none', padding: 0, background: 'transparent' }}
          aria-label="Information about the email"
        >
          <HelpIcon aria-hidden={true} />
        </button>
      </Tooltip>
      <a href="#" style={{ marginLeft: 'auto' }}>
        Details
      </a>
    </Grid>
    <FieldInput
      placeholder="example@domain.com"
      value={email}
      onChange={(e) => setEmail(e.target.value)}
      Action={
        <FieldAction aria-label="Show password">
          <Password aria-hidden={true} />
        </FieldAction>
      }
    />
    <FieldHint />
    <FieldError />
  </Stack>
</Field>
```

## With VoiceOver

Note for designers: the text field takes as much space as it's able to take. In our scenario, inside the application, this will be constrained by the grid (or any layout) mechanisms.

![Usage of the Field components using VoiceOver](https://user-images.githubusercontent.com/3874873/116384703-67421e80-a818-11eb-90fe-4b42d91dd719.gif)

rgesulfo reviewed (approved):
LGTM! :)

HichamELBSI commented on packages/strapi-design-system/src/Field/FieldAction.js:
```diff
@@ -0,0 +1,13 @@
+export const FieldAction = styled.button`
+  border: none;
+  background: transparent;
+  font-size: 1.6rem;
```
Can we add a "Don't forget to use the theme" please 🙏🏽

HichamELBSI commented on packages/strapi-design-system/src/Field/FieldAction.js:
```diff
@@ -0,0 +1,13 @@
+  border: none;
+  background: transparent;
+  font-size: 1.6rem;
+  margin-left: -100%;
```
I hate negative margin 😂  can you explain, why do you need it, please?

mfrachet replied:
The Input itself is wrapped in a parent container called "InputWrapper". This input wrapper has "flex" display. We can represent the big picture as following:

![Untitled-2021-03-27-2011](https://user-images.githubusercontent.com/3874873/116560877-16a6f000-a902-11eb-9d2c-63525993b4a4.png)

In order for the Icon to be inside the Input, it needs a way to be moved from the right edge. I know two ways to do that: make the InputWrapper relatively positioned, make the Action absolutely position, and use absolute coordinates or use a negative margin.

I honestly don't have a strong preference here.

Do you have a preferred way? Or even an option that I don't know about? 😊

mfrachet replied:
(InputWrapper is now a "row")

HichamELBSI replied:
TBH I prefer the `position relative` (I do it in the input UID), the margin is a bit weird when you inspect the dom (There is no impact in other elements). 

![Screenshot 2021-04-30 at 10 56 48](https://user-images.githubusercontent.com/7756284/116673217-67baf080-a9a3-11eb-95bd-c418935f6748.png)

mfrachet replied:
I took another approach and move the border to the parent. Then, the FieldAction only has to "push" things out . It fixes the negative margin issue and the input under the text issue 👍🏻

HichamELBSI commented on packages/strapi-design-system/src/Field/FieldInput.js:
```diff
@@ -0,0 +1,71 @@
+  color: ${({ theme }) => theme.colors.neutral800};
+  background: ${({ theme }) => theme.colors.neutral0};
+  font-weight: 400;
+  font-size: ${14 / 16}rem;
```
Same here, can you add a "Use the theme" todo comment for the font values please 😄

HichamELBSI commented on packages/strapi-design-system/src/Field/FieldInput.js:
```diff
@@ -0,0 +1,71 @@
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+`;
```
What do you think about creating something like a Flex component to avoid those CSS duplication?

mfrachet replied:
I'm pretty sure I can fix this with a grid 🤔

mfrachet replied:
Done with `Row` 😭

maevalienard commented:
- It seems the Details link doesn't follow the design rules to me, does it? I'm not sure if this is the right moment to mention it though, let me know if this specific part will be improved in another iteration :D 
<img width="943" alt="Capture d’écran 2021-04-29 à 16 39 19" src="https://user-images.githubusercontent.com/54801383/116569272-87053f80-a909-11eb-82c6-64a21459b1e0.png">
- Also related to this complex input, the icon color seems to be the wrong one. It looks like a neutral800 but should actually be a neutral500 (it's currently a neutral300 in the design but since we changed the other colors for an accessibility matters, we have to stay consistent). 

- The height of the input is currently 34px in Storybook but should be 40px -> The padding top and bottom currently applied are 8px each and should be at 12px (or rem sorry about that, I just inspected the element and got this information). 

- About the password input, we're supposed to have an eye icon on the right of the input. Does it appear? (I have two different plugin taking this space, so it's hard to say but I feel like it hasn't been implemented yet)

mfrachet commented:
(@maevalienard this is a perfect moment, thanks for the feedbacks, will take a look tomorrow 😊 )

mfrachet commented:
> It seems the Details link doesn't follow the design rules to me, does it? I'm not sure if this is the right moment to mention it though, let me know if this specific part will be improved in another iteration :D

The Link component has not been implemented yet. This is more a representation of the "layout" of a Field. This Field is not a "final" input, but it will aim to provide building blocks for other inputs 😊 . So, that's "normal" the link is not valid here

> - Also related to this complex input, the icon color seems to be the wrong one. It looks like a neutral800 but should actually be a neutral500 (it's currently a neutral300 in the design but since we changed the other colors for an accessibility matters, we have to stay consistent). 

Since we have not taken a strong decision on this icon, I've decided to put a "raw" icon in a "raw button". Again, it's a matter of layouts and positioning, not the style for this element exactly

> The height of the input is currently 34px in Storybook but should be 40px -> The padding top and bottom currently applied are 8px each and should be at 12px (or rem sorry about that, I just inspected the element and got this information).

I've moved to 12px 👍🏻 . Thanks for the review 😄 

> About the password input, we're supposed to have an eye icon on the right of the input. Does it appear? (I have two different plugin taking this space, so it's hard to say but I feel like it hasn't been implemented yet)

Same about the two first points :). This only aims to show what it may look like with an icon in it (layout). Also, it seems we don't have the "eye" icon in the strapi-icons package (or in figma?) and this is the reason why it's not this one in the example 😊 

When building a real text input with a password, the "eye" will be there 😊 

---

Also, it means that we can replace the "information" icon and the icon at the end of the input with any kind of other actions (for maybe, other field needs 😊 )

Thank you for the review 🙏🏻

maevalienard commented:
Looks amazing now, thanks for the changes @mfrachet and also thanks for all the explanations, that makes definitely sense! :D

soupette commented on packages/strapi-design-system/src/Field/FieldInput.js:
```diff
@@ -0,0 +1,67 @@
+FieldInput.displayName = 'FieldInput';
+
+FieldInput.defaultProps = {
+  Action: undefined,
```
Shouldn't it be null?

HichamELBSI reviewed (commented):
-

HichamELBSI reviewed (approved):
LGTM
